File: Week2/1 - Middle of the Linked List/Solution.py
#
# Created on Fri Apr 10 2020
#
# Title: Leetcode - Middle of the Linked List
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        p1 = p2 = head
        while p1 != None and p1.next != None:
            p1, p2 = p1.next.next, p2.next
        return p2

File: Week2/1 - Middle of the Linked List/Solution.cpp
/*
 * Created on Fri Apr 10 2020
 *
 * Title: Leetcode - Middle of the Linked List
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
    struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
      };
    
    public:
        ListNode* middleNode(ListNode* head) {
            if(head == NULL){
                return NULL;
            }
            
            ListNode *curr;
            curr = head;
            int count = 0;
            
            unordered_map<int, ListNode*> umap;
            while(curr != NULL) {
                umap[count] = curr; 
                curr = curr->next;
                count++;
            }
            
            int mid = count / 2;
            return umap[mid];
        }
};